### ISSUE TITLE
Add mkldnn related unit-tests

### ISSUE THREAD

**Issue by luotao1** [PaddlePaddle/Paddle#15615]
To improve the code (line+function) coverage, please **add mkldnn related unit-tests**:
http://ce.paddlepaddle.org:8080/viewLog.html?buildId=21110&buildTypeId=PaddlePaddleFramework_TestCoverage_TestCoverage&tab=report_project15_C___Coverage
- framework/ir/mkldnn
![image](https://user-images.githubusercontent.com/6836917/52030394-a22e4c80-2552-11e9-9328-29466a662d06.png)
- operators/elementwise/mkldnn
![image](https://user-images.githubusercontent.com/6836917/52032590-8fb91080-255c-11e9-91d4-a92950f18ced.png)
![image](https://user-images.githubusercontent.com/6836917/52032629-c727bd00-255c-11e9-9510-48872ec95503.png)
![image](https://user-images.githubusercontent.com/6836917/52031027-6e085b00-2555-11e9-9644-97fa91897c62.png)
![image](https://user-images.githubusercontent.com/6836917/52031039-7c567700-2555-11e9-8447-343850f02912.png)
![image](https://user-images.githubusercontent.com/6836917/52031049-8b3d2980-2555-11e9-92ae-bfeb2231c8b4.png)
![image](https://user-images.githubusercontent.com/6836917/52032662-dc045080-255c-11e9-96b0-a00438226f47.png)

**Comment by kbinias** [comment]
We reproduced your results with coverage testing tool.
- cmake .. -DWITH_GPU=OFF -DWITH_PROFILER=ON -DWITH_STYLE_CHECK=OFF -DWITH_MKLDNN=ON -DWITH_TESTING=ON -DWITH_COVERAGE=ON
- lcov --capture --directory . --output-file coverage.info
- genhtml coverage.info --output-directory out

We will add missing tests.

![image](https://user-images.githubusercontent.com/19502037/52485988-64689e00-2bba-11e9-8c1a-67703ffd723f.png)
![image](https://user-images.githubusercontent.com/19502037/52486012-6fbbc980-2bba-11e9-9ed4-dbdc59f3550e.png)
![image](https://user-images.githubusercontent.com/19502037/52486031-73e7e700-2bba-11e9-813b-66bf8de106e6.png)

**Comment by kbinias** [comment]
@luotao1 Whether it is possible to run more than 1 iteration in a single unit test for forward and backward phases ? We need to cover the part of the code responsible for reuse MKL-DNN primitives in backward (e.g. activation_mkldnn_op.cc::eltwise_grad function line 276-287). We have to run more than one iteration to enter this section of code.

**Comment by luotao1** [comment]
> Whether it is possible to run more than 1 iteration in a single unit test for forward and backward phases?

I think it's OK to run more than 1 iterations in elementwise_xxx_mkldnn unit-tests. However, if you add many more iterations, does the elasped time of unit-tests become larger?

**Comment by kbinias** [comment]
Not many, 2 iterations will be enough. Is there any unit test in PaddlePaddle which more than 1 iteration is started ?

**Comment by luotao1** [comment]
@kbinias There is no 2 iterations unitest using `OpTest`. Maybe you should implement yourselves, and refer to `test_batch_norm_op.py`.

**Comment by kbinias** [comment]
@luotao1 Thanks for you answer. What do you think about creating new test based on `test_imperative_resnet.py` with use_mkldnn equals to True ? It will cover batch_norm, conv, pool and activations.

**Comment by luotao1** [comment]
> What do you think about creating a new test based on `test_imperative_resnet.py` with use_mkldnn equals to True?

@kbinias I think it's OK.

**Comment by kbinias** [comment]
@luotao1 What do you think about closing this PR ? The problem with function coverage is IMO related to lcov tool.

![image](https://user-images.githubusercontent.com/19502037/53946808-cbe41180-40c4-11e9-94db-574fb6a2a34f.png)
![image](https://user-images.githubusercontent.com/19502037/53946820-d1415c00-40c4-11e9-8bab-9bcaed8accbe.png)

**Comment by luotao1** [comment]
The Functions coverage is less than 80% in `operators/mkldnn`. You can add the unit-tests for `activation_mkldnn_op.cc`, `conv_mkldnn_op.cc` and `transpose_mkldnn_op.cc` when you are convenient. @kbinias

**Comment by kbinias** [comment]
@luotao1  Coverage feature in GCC doesn't support virtual destructors correctly. There is no way to cover tests in 100% if test uses Inheritance. You can find out more details below with good explanation:
https://stackoverflow.com/questions/25662174/is-there-a-way-to-call-the-deleting-destructor-of-a-pure-virtual-class.

One of potential workaround might be post-processing and remove all destructors from tests, we could consider that it’s a noise. The destructors should not be a part of final calculation.

**Comment by luotao1** [comment]
Got it. Both @tensor-tang and I think destructors are normal. 
However, the current coverage by May 12th is http://ce.paddlepaddle.org:8080/viewLog.html?buildId=31464&buildTypeId=PaddlePaddleFramework_TestCoverage_TestCoverage&tab=report_project15_C___Coverage
![image](https://user-images.githubusercontent.com/6836917/54284172-70d97f80-45da-11e9-875f-bd9d4040222b.png)
Why the coverage of `batch_norm_mkldnn_op` and `lrn_mkldnn_op` cut down that before?

**Comment by lidanqing-vv** [comment]
我们的结果完全不一样，lrn应该是100%的，您可以用下面的编译指令重新生成一下吗  @luotao1 谢谢
⦁        cmake .. -DWITH_GPU=OFF -DWITH_PROFILER=ON -DWITH_STYLE_CHECK=OFF -DWITH_MKLDNN=ON -DWITH_TESTING=ON -DWITH_COVERAGE=ON 
⦁        make -j 12
⦁        ctest -R mkldnn
⦁        lcov --capture --directory . --output-file coverage.info 
⦁        genhtml --demangle-cpp coverage.info --output-directory out

**Comment by luotao1** [comment]
我是CI上的结果。

**Comment by lidanqing-vv** [comment]
Could you please tell how to reproduce it with commands ? It seems we have different procedure... 请问怎么reproduce呢，可以有什么指令吗？ 非常感谢！

**Comment by lidanqing-vv** [comment]
We are checking build log, hope we could find something. Thanks. If you have any further information, please inform us. Thanks

**Comment by pawelpiotrowicz** [comment]
Macro REGISTER_OP_KERNEL is not able to set correct handle for all grad functions such TouchOpKernelRegistrar_transpose_grad_MKLDNN_DEFAULT_TYPE().
Due to missed registration, we can’t cover all functions in tests. The issue is globally, moreover practical working example how to register kernel and enforce touchopkernelregistrar_grad correctly doesn’t exist. The open question is - do we really should care about it?

**Comment by kbinias** [comment]
@luotao1 What do you think about above problem ?

**Comment by panyx0718** [comment]
do you mean paddle can't register some kernel or paddle can't select some registered kernels?

can you search REGISTER_OP_KERNEL_WITH_CUSTOM_TYPE and see if it works for you?

**Comment by luotao1** [comment]
@shanyi15 Could you help to update [How to add new op](http://paddlepaddle.org/documentation/docs/en/1.3/advanced_usage/development/new_op/new_op_en.html) to introduce so many different REGISTER MACRO?

**Comment by danqing-li** [comment]
@luotao1  About batch_norm_mkldnn and lrn_mkldnn line coverages, I have some idea and made some fix locally. I could build and make with docker[latest-dev],  but I don't know what is the building and testing script. **Something like ./paddle_build_cov.sh**. I also don't know what **main cov** means in the build log on TeamCity. Could you share build script with me? This will speed up the work. Thank you!

**Comment by luotao1** [comment]
@lidanqing-intel The different between `paddle_build_cov.sh` adn `paddle_build.sh` is only in `function main() `
```
cov)
        cmake_gen ${PYTHON_ABI:-""}
        build
        run_test
        ;;
```
@kolinwei will update `paddle_build.sh` in next PR.

**Comment by lidanqing-vv** [comment]
> @lidanqing-intel The different between `paddle_build_cov.sh` adn `paddle_build.sh` is only in `function main() `
> 
> ```
> cov)
>         cmake_gen ${PYTHON_ABI:-""}
>         build
>         run_test
>         ;;
> ```
> @kolinwei will update `paddle_build.sh` in next PR.

Thank you!

**Comment by luotao1** [comment]
please see #16324

**Comment by pawelpiotrowicz** [comment]
@panyx0718 @luotao1 @kbinias 


The point it that Register_Op_kernel consist of Register_op_Kernel_with_custom_type.
```
#define REGISTER_OP_KERNEL(op_type, library_type, place_class, __VA_ARGS__)
REGISTER_OP_KERNEL_WITH_CUSTOM_TYPE( op_type, library_type, place_class, DEFAULT_TYPE, ::paddle::framework::OpKernelType::kDefaultCustomizedTypeValue, __VA_ARGS__)
```

To better understand what I’m trying to tell is a simple example below.
We have two execution on macro ( for instance transpose_mkldnn_op.cc) .
```
REGISTER_OP_KERNEL(transpose, MKLDNN, ::paddle::platform::CPUPlace,
                   ops::TransposeMKLDNNOpKernel<float>);
REGISTER_OP_KERNEL(transpose_grad, MKLDNN, ::paddle::platform::CPUPlace,
                   ops::TransposeMKLDNNGradOpKernel<float>);                   
```


So we got two functions, only first one will be executed – I expect both of them. 
```
int TouchOpKernelRegistrar_transpose_MKLDNN_DEFAULT_TYPE() { __op_kernel_registrar_transpose_MKLDNN_DEFAULT_TYPE__ .Touch(); return 0; }
int TouchOpKernelRegistrar_transpose_grad_MKLDNN_DEFAULT_TYPE() { __op_kernel_registrar_transpose_grad_MKLDNN_DEFAULT_TYPE__ .Touch(); return 0; }
```

**Comment by lidanqing-vv** [comment]
Hi, @luotao1  Both batch_norm_mkldnn and lrn_mkldnn failed because of timeout and it passed with good line coverage in our machines, without docker and tests are run in serial.

I just update the progress with you. And similar problems happened before so I put comment here.

I consider to set in **cmake/generic.cmake**, in **line 394 and line 721**
**set_tests_properties(${TARGET_NAME} PROPERTIES TIMEOUT 1000)**

We also found the unit tests are run in parallel, while timeout limit is set 600 strictly. We consider setting mkldnn unit test to run in the serial way as in PR #16233 .

**Comment by luotao1** [comment]
@lidanqing-intel GOT it, Thanks very much for your analysis! 
> it passed with good line coverage in our machines

Could you give the line coverage here?

**Comment by lidanqing-vv** [comment]
As in [krzysztof's comment](https://github.com/PaddlePaddle/Paddle/issues/15615#issuecomment-470462749).
line cov of batch_norm_mkldnn_op.cc 91%
line cov of lrn_mkldnn_op.cc 100%

**Comment by luotao1** [comment]
@kbinias I close this issue due to https://github.com/PaddlePaddle/Paddle/issues/15615#issuecomment-474862344
@pawelpiotrowicz For https://github.com/PaddlePaddle/Paddle/issues/15615#issuecomment-474815683, how about creating a new issue to discuss it?
